### PR TITLE
email: do not send certain emails to unverified emails

### PR DESCRIPTION
**Pull request by bobheadxi**
We should seek to avoid damaging email reputations by validating if emails are verified before sending certain types of potentially high-frequency, non-essential emails to user primary emails, in this change:

- Account update notifications
- Code monitors

In practice, at the moment only builtin auth signups will create accounts that have unverified primary emails. External service accounts are created with primary emails already verified, and so are those created with the `createUser` mutation (which I am hoping to change in https://github.com/sourcegraph/sourcegraph/pull/46187).

Today, a user could create a user, which automatically marks the email as verified, and then proceed to set up a code monitor that accidentally sends large numbers of emails that all bounce, which has implications for Cloud's managed SMTP services.

The risk: there was at least one incident recently where a small number of code monitors issued over a thousand emails in a few days due to an unforeseen issue with retry handling (https://github.com/sourcegraph/sourcegraph/pull/45641), and at least [one instance of a user creating an account with an invalid email](https://sourcegraph.slack.com/archives/C03LCPCT3SP/p1672943730325969).

Stacked on https://github.com/sourcegraph/sourcegraph/pull/46183

## Test plan

image

### CODE REVIEW

**Comment by bobheadxi** [comment]
Current dependencies on/for this PR:
* main
  * **PR #46183** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/46183" target="_blank">Graphite</a> 
    * **PR #46184** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/46184" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/46184?utm_source=stack-comment).

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (+0.04 kb) | 0.00% (+0.63 kb) | 0.01% (+0.59 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/01-05-email_do_not_send_emails_to_unverified_emails/compare-report.html;tab=live_object) for a full comparison between the commits 5b40d8f and 7a6dc29 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: cmd/frontend/backend/user_emails.go [hunk 1]
```diff
@@ -293,16 +293,16 @@ func (e *userEmails) ResendVerificationEmail(ctx context.Context, userID int32,
...
+		return errors.Wrap(err, "get user primary email")
+	}
+	if !verified {
+		return errors.Newf("unable to send email to user ID %d's unverified primary email address", id)
 	}
 	usr, err := e.db.Users().GetByID(ctx, id)
 	if err != nil {
-		logger.Warn("Failed to get user from database", log.Error(err))
-		return err
+		return errors.Wrap(err, "get user")
```
Comment (unknwon) [inline cmd/frontend/backend/user_emails.go:305]: 👍

File: enterprise/cmd/frontend/internal/auth/githuboauth/session.go [hunk 2]
```diff
@@ -111,11 +111,15 @@ func (s *sessionIssuerHelper) GetOrCreateUser(ctx context.Context, token *oauth2
 	for i, attempt := range attempts {
 		userID, safeErrMsg, err := auth.GetAndSaveUser(ctx, s.db, auth.GetAndSaveUserOp{
 			UserProps: database.NewUser{
-				Username:        login,
+				Username: login,
+
+				// This email is coming from an external source, so we can safely assume
+				// that the email is verified elsewhere.
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/auth/githuboauth/session.go:117]: nit: I think this is not 100% accurate
```suggestion
				// We always only take verified emails from an external source.
```

see line 64 of this file.

File: enterprise/cmd/frontend/internal/auth/httpheader/middleware.go [hunk 3]
```diff
@@ -92,7 +92,13 @@ func middleware(db database.DB) func(next http.Handler) http.Handler {
 				}
 			}
 			userID, safeErrMsg, err := auth.GetAndSaveUser(r.Context(), db, auth.GetAndSaveUserOp{
-				UserProps: database.NewUser{Username: username, Email: rawEmail, EmailIsVerified: true},
+				UserProps: database.NewUser{
+					Username: username,
+					Email:    rawEmail,
+					// This email is coming from an external source, so we can safely assume
+					// that the email is verified elsewhere.
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/auth/httpheader/middleware.go:99]: Ditto

**Review by unknwon — APPROVED** [review]
Approve to unblock, just one minor issue

**Review by limitedmage — APPROVED** [review]
`EmailAction.tsx` looks good! Didn't review the Go code.

**Review by evict — APPROVED** [review]

**Pull request merged**